app/controllers/reviews_controller.rb

```diff
@@ -20,11 +20,35 @@ def create
   end
 
   def edit
+    @review = Review.find(params[:id])
+    @piece = Piece.find(params[:piece_id])
+    if current_user != @review.user && !current_user.admin
+      flash[:notice] = "You can only edit a review you created"
+      redirect_to piece_path(@piece)
+    end
   end
 
   def update
+    @review = Review.find(params[:id])
+    @piece = Piece.find(params[:piece_id])
+    if @review.update_attributes(review_params)
+      flash[:notice] = "Review edited successfully"
+      redirect_to piece_path(@piece)
+    else
+      flash[:errors] = @review.errors.full_messages.join(". ")
+      render :edit
+    end
   end
 
+  def destroy
+    review = Review.find(params[:id])
+    piece = Piece.find(params[:piece_id])
+    review.destroy
+    flash[:success] = "Review Deleted"
+    redirect_to piece_path(piece)
+  end
+
+
```

Comment:
Consider adding a `private` method to this controller to help DRY up your code:

```
class ReviewsController < ApplicationController
  ...

  def edit
    review
    piece
    ...
  end

  def update
    piece
    if review.update_attributes(review_params)
      ...
    else
      ...
    end
    ...
  end

  def destroy
    piece
    review.destroy
    ...
  end

  private

  def review
    @review ||= Review.find(params[:id])
  end

  def piece
    @piece ||= Piece.find(params[:piece_id])
  end
end
```

This allows you to better isolate your ReviewsController from other classes, so that they only interact through that single method, but it also helps to shorten controller actions (notice that you don't have to have two separate lines to find the review and destroy it, for example?), as you can use these accessor methods to either find OR return an already located resource